src/plots/cartesian/axes.js

```diff
@@ -657,8 +657,36 @@ function isClose(a, b) {
 axes.prepTicks = function(ax, opts) {
     var rng = Lib.simpleMap(ax.range, ax.r2l, undefined, undefined, opts);
 
-    // calculate max number of (auto) ticks to display based on plot size
-    if(ax.tickmode === 'auto' || !ax.dtick) {
+    // sync ticks with the overlaying defined axis
+    if(ax.tickmode === 'sync') {
```

Comment:
I suggest we move this logic into a function named `syncTicks`  and adjust it to operate on every tick of the main axis instead of start & end only.
That way we don't have to compute `dtick` and `tick0` and it is going to work similar to when `tickmode` is set to "array". 
So similarly you could call `syncTicks` above/below this block. 
https://github.com/plotly/plotly.js/blob/6352d93c00db2b98b7acf4756ef8743bd9631960/src/plots/cartesian/axes.js#L971